platform/ios/src/MGLScaleBarView.m

```diff
@@ -0,0 +1,387 @@
+    if (self.isVisible) {
+        return;
+    }
+
+    self.visible = YES;
+    
+    [UIView animateWithDuration:.2 delay:0 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 1.0;
+    } completion:nil];
+}
+
+- (void)fadeOut {
+    if (!self.isVisible) {
+        return;
+    }
+    
+    self.visible = NO;
+    
+    [UIView animateWithDuration:.3 delay:0.7 options:UIViewAnimationOptionBeginFromCurrentState animations:^{
+        self.alpha = 0.0;
+    } completion:nil];
+}
+
+#pragma mark - Setters
+
+- (void)setMetersPerPoint:(CLLocationDistance)metersPerPoint {
+    if (_metersPerPoint == metersPerPoint) {
+        return;
+    }
+    
+    [self fadeIn];
+    
+    _metersPerPoint = metersPerPoint;
+    self.row = [self preferredRowInRows:[self validRows]];
+    
+    [self invalidateIntrinsicContentSize];
+    [self setNeedsLayout];
+}
+
+- (void)setRow:(NSArray<NSNumber *> *)row {
+    if (![row isEqualToArray:_row]) {
+        [_bars makeObjectsPerformSelector:@selector(removeFromSuperview)];
+        [_labels makeObjectsPerformSelector:@selector(removeFromSuperview)];
+        _bars = nil;
+        _labels = nil;
+        _row = row;
+    }
+}
+
+#pragma mark - Views
+
+- (NSArray<UIView *> *)bars {
+    if (!_bars) {
+        NSMutableArray *bars = [NSMutableArray array];
+        for (NSUInteger i = 0; i < self.row.lastObject.integerValue; i++) {
+            UIView *bar = [[UIView alloc] init];
+            [bars addObject:bar];
+            [self.containerView addSubview:bar];
+        }
+        _bars = bars;
+    }
+    return _bars;
+}
+
+- (NSArray<UILabel *> *)labels {
+    if (!_labels) {
+        NSMutableArray *labels = [NSMutableArray array];
+        for (NSUInteger i = 0; i <= self.row.lastObject.integerValue; i++) {
+            UILabel *label = [[MGLScaleBarLabel alloc] init];
+            label.font = [UIFont systemFontOfSize:8 weight:UIFontWeightMedium];
+            NSCharacterSet *characterSet = [[NSCharacterSet decimalDigitCharacterSet] invertedSet];
+            label.text = [[[self.formatter stringFromDistance:0] componentsSeparatedByCharactersInSet:characterSet] componentsJoinedByString:@""];
+            [label setNeedsDisplay];
+            [label sizeToFit];
+            [labels addObject:label];
+            [self addSubview:label];
+        }
+        _labels = labels;
+    }
+    return _labels;
+}
+
+#pragma mark - Layout
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+    
+    if (!self.row.lastObject || !self.row.lastObject.integerValue) {
+        // Current distance is not within allowed range
+        return;
+    }
+    
+    [self updateLabels];
+    [self layoutBars];
+    [self layoutLabels];
+}
+
+- (void)layoutBars {
+    CGFloat barWidth = (CGRectGetWidth(self.bounds) - self.borderWidth * 2.0f) / self.bars.count;
+    
+    NSUInteger i = 0;
+    for (UIView *bar in self.bars) {
+        CGFloat xPosition = barWidth * i + self.borderWidth;
```

Comment:
The compass already appears in the top-left corner due to Auto Layout mirroring. Ideally everything about the scale bar would be mirrored, so that it reads from right to left.